sql/types/datetime.go

```diff
@@ -361,16 +363,40 @@ func (t datetimeType) ConvertWithoutRangeCheck(ctx context.Context, v interface{
 		res = res.Truncate(24 * time.Hour)
 	}
 
-	return res, nil
+	return res, err
+}
+
+func (t datetimeType) parseDatetime(value string) (time.Time, bool, error) {
+	if t, err := time.Parse(TimezoneTimestampDatetimeLayout, value); err == nil {
+		return t.UTC(), true, nil
+	}
+
+	valueLen := len(value)
+	end := valueLen
+
+	for end > 0 {
```

Comment:
I believe the shortest possible valid date is `2000-1-1`, so you can bail a little earlier for completely invalid dates.